README.md
# MDM Agent for Android

![Flyve MDM banner](https://user-images.githubusercontent.com/663460/26935464-54267e9c-4c6c-11e7-86df-8cfa6658133e.png)

[![License](https://img.shields.io/github/license/flyve-mdm/android-mdm-agent.svg?&label=License)](https://github.com/flyve-mdm/android-mdm-agent/blob/develop/LICENSE.md)
[![Follow twitter](https://img.shields.io/twitter/follow/FlyveMDM.svg?style=social&label=Twitter&style=flat-square)](https://twitter.com/FlyveMDM)
[![Project Status: WIP](http://www.repostatus.org/badges/latest/wip.svg)](http://www.repostatus.org/)
[![Telegram Community](https://img.shields.io/badge/Telegram-Community-blue.svg)](https://t.me/flyvemdm)
[![IRC Chat](https://img.shields.io/badge/IRC-%23flyvemdm-green.svg)](http://webchat.freenode.net/?channels=flyve-mdm)
[![Conventional Commits](https://img.shields.io/badge/Conventional%20Commits-1.0.0-yellow.svg)](https://conventionalcommits.org)
[![Greenkeeper badge](https://badges.greenkeeper.io/flyve-mdm/android-mdm-agent.svg)](https://greenkeeper.io/)
[![GitHub release](https://img.shields.io/github/release/flyve-mdm/android-mdm-agent.svg)](https://github.com/flyve-mdm/android-mdm-agent/releases)

Flyve MDM is a Mobile device management software that enables you to secure and manage all the mobile devices of your business or family via a web-based console.

To get started, check out [Flyve MDM Website](https://flyve-mdm.com/)!

## Table of contents

* [Synopsis](#synopsis)
* [Build Status](#build-status)
* [Compatibility Matrix](#compatibility-matrix)
* [Installation](#installation)
* [Documentation](#documentation)
* [Versioning](#versioning)
* [Contact](#contact)
* [Professional Services](#professional-services)
* [Contribute](#contribute)
* [Copying](#copying)

## Synopsis

The Android Agent will manage all your mobile devices effectively, including a range of powerful security features to protect sensitive company data stored on mobile devices, allowing you to manage your mobile fleet security policy with precision.

The Android Agent will help you to:

* Configure and deploy your mobile fleet from the web-based administration console
* Install and uninstall remote applications
* Send files remotely to all terminals
* Deploy and control Bluetooth and Wi-Fi connectivity
* Activate real-time geolocation of your terminals
* Delete, partially or totally, your data in case of loss or theft
* Create an inventory of all of your company terminals
* Keep control and protect yourself from cyber-attacks:
  * Set the level of complexity of your passwords
  * Activate mobile device encryption
  * Lock the mobile device remotely
  * Control the authorisation for the use of the camera
  * Erase all data from the remote terminal (reset)

For more information visit our [Official Website](http://flyve.org/android-mdm-agent/).

## Build Status

| **LTS** | **Bleeding Edge** |
|:---:|:---:|
| [![Build Status](https://circleci.com/gh/flyve-mdm/android-mdm-agent/tree/master.svg?style=svg)](https://circleci.com/gh/flyve-mdm/android-mdm-agent/tree/master) | [![Build Status](https://circleci.com/gh/flyve-mdm/android-mdm-agent/tree/develop.svg?style=svg)](https://circleci.com/gh/flyve-mdm/android-mdm-agent/tree/develop) |

## Compatibility Matrix

### Flyve MDM projects

|MDM Agent|0.99.x|1.0.0|2.0.0|
|:---|---|---|---|
|GLPI|9.1|9.2|9.2|
|Flyve MDM plugin|-|2.0.0|2.0.0|
|Web MDM Dashboard|-|2.0.0|2.0.0|

### MDM Agent & Android

|API|Android|Support|SSL|
|:---:|:---:|:---:|:---:|
| 16 | Android 4.1 | X | - |
| 17 | Android 4.2 | X | - |
| 18 | Android 4.3 | X | - |
| 19 | Android 4.4 | X | - |
| 20 | Android 4.4 | X | - |
| 21 | Android 5.0 | X | X |
| 22 | Android 5.1 | X | X |
| 23 | Android 6.0 | X | X |
| 24 | Android 7.0 | X | X |
| 25 | Android 7.1.1 | X | X |
| 26 | Android 8.0 | - | - |

## Installation

[<img src="https://user-images.githubusercontent.com/663460/26973322-4ddf78a4-4d16-11e7-8b58-4c03b4bc2490.png" alt="Get it on Google Play" height="60">](https://play.google.com/store/apps/details?id=org.flyve.mdm.agent) [<img src="https://f-droid.org/badge/get-it-on.png" alt="Get it on F-Droid" height="60">](https://f-droid.org/app/org.flyve.mdm.agent) [<img src="https://user-images.githubusercontent.com/663460/26973090-f8fdc986-4d14-11e7-995a-e7c5e79ed925.png" alt="Download APK from GitHub" height="60">](https://github.com/flyve-mdm/android-mdm-agent/releases/latest)

### As System App

Follow the steps in this article to learn [How to Install as System App](http://flyve.org/android-mdm-agent/howtos/installation).

### Beta Testing App

* Install the [Beta version of the app](https://play.google.com/apps/testing/org.flyve.mdm.agent)

## Documentation

We maintain a detailed documentation of the project on the website, check the [How-tos](http://flyve.org/android-mdm-agent/) and [Development](http://flyve.org/android-mdm-agent/howtos/) section.

## Versioning

In order to provide transparency on our release cycle and to maintain backward compatibility, Flyve MDM is maintained under [the Semantic Versioning guidelines](http://semver.org/). We are committed to following and complying with the rules, the best we can.

See [the tags section of our GitHub project](https://github.com/flyve-mdm/android-mdm-agent/tags) for changelogs for each release version of Flyve MDM. Release announcement posts on [the official Teclib' blog](http://www.teclib-edition.com/en/communities/blog-posts/) contain summaries of the most noteworthy changes made in each release.

## Contact

For notices about major changes and general discussion of Flyve MDM development, subscribe to the [/r/FlyveMDM](http://www.reddit.com/r/FlyveMDM) subreddit.
You can also chat with us via IRC in [#flyve-mdm on freenode](http://webchat.freenode.net/?channels=flyve-mdm) or [@flyvemdm on Telegram](https://t.me/flyvemdm). Ping me @rafaelje if you get stuck.

## Professional Services

The Flyve MDM and GLPI Network services are available through our [Partner's Network](http://www.teclib-edition.com/en/partners/). We provide special training, bug fixes with editor subscription, contributions for new features, and more.

Obtain a personalized service experience, associated with benefits and opportunities.

## Contribute

Want to file a bug, contribute some code, or improve documentation? Excellent! Read up on our guidelines for [contributing](./CONTRIBUTING.md) and then check out one of our issues in the [Issues Dashboard](https://github.com/flyve-mdm/android-mdm-agent/issues).

## Copying

* **Name**: [Flyve MDM](https://flyve-mdm.com/) is a registered trademark of [Teclib'](http://www.teclib-edition.com/en/).
* **Code**: you can redistribute it and/or modify it under the terms of the GNU General Public License ([GPLv3](https://www.gnu.org/licenses/gpl-3.0.en.html)).
* **Documentation**: released under Attribution 4.0 International ([CC BY 4.0](https://creativecommons.org/licenses/by/4.0/)).


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

<a name="2.0.0-rc.6"></a>
# [2.0.0-rc.6](https://github.com/flyve-mdm/android-mdm-agent/compare/2.0.0-rc.5...2.0.0-rc.6) (2018-12-05)


### Bug Fixes

* **connection:** add session token ([c1cd8f0](https://github.com/flyve-mdm/android-mdm-agent/commit/c1cd8f0))
* **connection:** add user token on initsession ([2794374](https://github.com/flyve-mdm/android-mdm-agent/commit/2794374))
* **download:** add session token on header ([126f26d](https://github.com/flyve-mdm/android-mdm-agent/commit/126f26d))
* **enroll:** remove mqtt restriction t to save ([04c747a](https://github.com/flyve-mdm/android-mdm-agent/commit/04c747a))
* **log:** add connection and reconnection logs ([f9ec199](https://github.com/flyve-mdm/android-mdm-agent/commit/f9ec199))
* **log:** add date and keep clear the log ([e9a6e99](https://github.com/flyve-mdm/android-mdm-agent/commit/e9a6e99))
* **log:** add limit on the log file text with 3MB ([1b1ea98](https://github.com/flyve-mdm/android-mdm-agent/commit/1b1ea98))
* **log:** add log data structure ([c7dcfcd](https://github.com/flyve-mdm/android-mdm-agent/commit/c7dcfcd))
* **log:** check the message value ([dc40d93](https://github.com/flyve-mdm/android-mdm-agent/commit/dc40d93))
* **log:** clear the data with ui ([188bf3c](https://github.com/flyve-mdm/android-mdm-agent/commit/188bf3c))
* **log:** implement log with sqllite ([7ef05bb](https://github.com/flyve-mdm/android-mdm-agent/commit/7ef05bb))
* **log:** implement where  on log ([ea6013a](https://github.com/flyve-mdm/android-mdm-agent/commit/ea6013a))
* **log:** load logs from sqllite ([0c86500](https://github.com/flyve-mdm/android-mdm-agent/commit/0c86500))
* **log:** refactor the save log ([ffdd295](https://github.com/flyve-mdm/android-mdm-agent/commit/ffdd295))
* **mqtt:** add certification ([318b619](https://github.com/flyve-mdm/android-mdm-agent/commit/318b619))
* **mqtt:** add TrustManagerFactory on  the mqtt options ([915cd60](https://github.com/flyve-mdm/android-mdm-agent/commit/915cd60))
* **mqtt:** fix the FlyvemdmManifest topic ([bacaf17](https://github.com/flyve-mdm/android-mdm-agent/commit/bacaf17))
* **mqtt:** remove increase tryEverySeconds ([a7d9267](https://github.com/flyve-mdm/android-mdm-agent/commit/a7d9267))
* **mqtt:** remove the retained true ([327586d](https://github.com/flyve-mdm/android-mdm-agent/commit/327586d))
* **mqtt:** replace the TLSv1.2 for TLS ([fede198](https://github.com/flyve-mdm/android-mdm-agent/commit/fede198))
* **mqtt:** restore increase seconds ([3a893c2](https://github.com/flyve-mdm/android-mdm-agent/commit/3a893c2))
* **mqtt:** rewords variables ([b2f6f7c](https://github.com/flyve-mdm/android-mdm-agent/commit/b2f6f7c))
* **mqtt:** send ? when gps is not available ([316ea15](https://github.com/flyve-mdm/android-mdm-agent/commit/316ea15))
* **mqtt:** send online status ([073bc2d](https://github.com/flyve-mdm/android-mdm-agent/commit/073bc2d))
* **mqtt:** send the status connected ([27ce044](https://github.com/flyve-mdm/android-mdm-agent/commit/27ce044))
* **mqtt:** set reconnection counter to zero ([5ae89ff](https://github.com/flyve-mdm/android-mdm-agent/commit/5ae89ff))
* **mqtt:** set status true if is  connected ([9a83eb1](https://github.com/flyve-mdm/android-mdm-agent/commit/9a83eb1))
* **mqtt:** validate unique mqtt client ([b0a2936](https://github.com/flyve-mdm/android-mdm-agent/commit/b0a2936))
* **MqttModel:** fix will for online status ([0eb4276](https://github.com/flyve-mdm/android-mdm-agent/commit/0eb4276))
* **MqttModel:** offline will not on the good topic ([791a5cb](https://github.com/flyve-mdm/android-mdm-agent/commit/791a5cb))
* **policy:** remove lock  var on thread manager ([38f9bd9](https://github.com/flyve-mdm/android-mdm-agent/commit/38f9bd9))
* **policy:** send ? character iif gps fail ([236fe67](https://github.com/flyve-mdm/android-mdm-agent/commit/236fe67))
* **policy:** update the log  message ([9c13ff6](https://github.com/flyve-mdm/android-mdm-agent/commit/9c13ff6))
* **policy:** use the right policies connectiviity ([a9fe7cb](https://github.com/flyve-mdm/android-mdm-agent/commit/a9fe7cb))
* **service:** remove stop service from ui ([0cc9b27](https://github.com/flyve-mdm/android-mdm-agent/commit/0cc9b27))


### Features

* **deeplink:** add  the textview information ([f59b0c9](https://github.com/flyve-mdm/android-mdm-agent/commit/f59b0c9))
* **deeplink:** show the deeplink infomrmation ([0546ae2](https://github.com/flyve-mdm/android-mdm-agent/commit/0546ae2))
* **policies:** add policies module ([625ceca](https://github.com/flyve-mdm/android-mdm-agent/commit/625ceca))
* **wipe:** set external storage flag on wipe ([#623](https://github.com/flyve-mdm/android-mdm-agent/issues/623)) ([33c6dd2](https://github.com/flyve-mdm/android-mdm-agent/commit/33c6dd2))



<a name="2.0.0-rc.5"></a>
# [2.0.0-rc.5](https://github.com/flyve-mdm/android-mdm-agent/compare/2.0.0-rc.4...2.0.0-rc.5) (2018-10-03)


### Bug Fixes

* **app:** check application package ([26a8285](https://github.com/flyve-mdm/android-mdm-agent/commit/26a8285))
* **app:** check application package ([8fc2f08](https://github.com/flyve-mdm/android-mdm-agent/commit/8fc2f08))
* **app:** implement version code and name ([8b6b175](https://github.com/flyve-mdm/android-mdm-agent/commit/8b6b175))
* **application:** improve the application status ([20c2f90](https://github.com/flyve-mdm/android-mdm-agent/commit/20c2f90))
* **apps:** add thread manager ([802c0a1](https://github.com/flyve-mdm/android-mdm-agent/commit/802c0a1))
* **apps:** add thread manager ([ce00727](https://github.com/flyve-mdm/android-mdm-agent/commit/ce00727))
* **apps:** implement thread manager ([abb3f3a](https://github.com/flyve-mdm/android-mdm-agent/commit/abb3f3a))

ci/deploy.sh
#!/bin/bash

#   Copyright © 2017 Teclib. All rights reserved.
#
#   This file is part of flyve-mdm-android-agent
#
# flyve-mdm-android-agent is a subproject of Flyve MDM. Flyve MDM is a mobile
# device management software.
#
# Flyve MDM is free software: you can redistribute it and/or
# modify it under the terms of the GNU General Public License
# as published by the Free Software Foundation; either version 3
# of the License, or (at your option) any later version.
#
# Flyve MDM is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
# ------------------------------------------------------------------------------
# @author    Rafael Hernandez - rafaelje
# @date      24/9/17
# @copyright Copyright © 2017 Teclib. All rights reserved.
# @license   GPLv3 https://www.gnu.org/licenses/gpl-3.0.html
# @link      https://github.com/flyve-mdm/flyve-mdm-android-agent
# @link      https://flyve-mdm.com
# ------------------------------------------------------------------------------

# create enviroment vars to work with fastlane
echo TELEGRAM_WEBHOOKS=$TELEGRAM_WEBHOOKS > .env
echo GIT_REPO=$TRAVIS_REPO_SLUG >> .env
echo GIT_BRANCH=$TRAVIS_BRANCH >> .env

# Store information on gh-pages branch
# - Create documentation with javadoc
# - Create Coverage Reporte
# - Create Changelog
# - Push this file to gh-pages branch
updateGHPAGES() {
    # Generate javadoc this folder must be on .gitignore
    javadoc -d ./reports$1/javadoc -sourcepath ./app/src/main/java -subpackages .

    # create code coverage report
    gradle createDebugCoverageReport

    # get gh-pages branch
    git fetch origin gh-pages

    # move to gh-pages
    git checkout gh-pages

    # add javadoc folder
    git add reports$1/javadoc

    # create commit for documentation
    git commit -m "docs(javadoc): update javadoc with version ${GIT_TAG}"

    # move code coverage
    mv -v app/build/reports/coverage reports$1

    #move Android test
    mv -v app/build/reports/androidTests reports$1

    # replace .resources with resource because github don't support folders with "_" or "." at the beginning
    mv reports$1/coverage/debug/.resources reports$1/coverage/debug/resources

    index=$(<reports$1/coverage/debug/index.html)
    newindex="${index//.resources/resources}"
    echo $newindex > reports$1/coverage/debug/index.html

    # add code coverage
    git add reports$1/coverage

    # add Android Tests
    git add reports$1/androidTests

    # create commit
    git commit -m "docs(coverage): update code coverage with version ${GIT_TAG}"

    # clean unstage file on gh-pages to remove all others files gets on checkout
    git clean -fdx

    # remove CHANGELOG.md
    rm CHANGELOG.md

    # get changelog from branch
    git checkout $TRAVIS_BRANCH CHANGELOG.md

    # Create header content to work with gh-pages templates
    HEADER="---"$'\r'"layout: modal"$'\r'"title: changelog"$'\r'"---"$'\r\r'

    # Duplicate CHANGELOG.md
    cp CHANGELOG.md CHANGELOG_COPY.md

    # Add header to CHANGELOG.md
    (echo $HEADER ; cat CHANGELOG_COPY.md) > CHANGELOG.md

    # Remove CHANGELOG_COPY.md
    rm CHANGELOG_COPY.md

    # add
    git add CHANGELOG.md

    # create commit
    git commit -m "docs(changelog): update changelog$1 with version ${GIT_TAG}"

    # push to branch
    git push origin gh-pages
}

# Configure git with Flyve bot user to make commits on repositories
# and add new origin with Flyve bot credentials
configGit() {
    # config git
    git config --global user.email $GH_EMAIL
    git config --global user.name "Flyve MDM"
    git remote remove origin
    git remote add origin https://$GH_USER:$GH_TOKEN@github.com/$TRAVIS_REPO_SLUG.git
}

#-----------------------------------------------------------------
# DEVELOP DEPLOY
# - send to google play like beta
#-----------------------------------------------------------------
if [[ "$TRAVIS_BRANCH" == "develop" && "$TRAVIS_PULL_REQUEST" == "false" && "$TRAVIS_RUN" == "true" ]]; then

    # decrypt deploy on google play file
    openssl aes-256-cbc -K $encrypted_27dcfd0dda78_key -iv $encrypted_27dcfd0dda78_iv -in gplay.tar.gz.enc -out ci/gplay.tar.gz -d

    # uncompress cert file
    tar -zxvf ci/gplay.tar.gz -C ci/

    # sign and deploy to store with fastlane
    fastlane android beta storepass:'$KEYSTORE' keypass:'$ALIAS'

    # config git
    configGit

    #------------------------ UPDATE CHANGES --------------------------
    git add -u
    git commit -m "ci(build): release **beta** for version $GIT_TAG-beta"

    git push origin $TRAVIS_BRANCH

    #------------------------ GH-PAGES --------------------------

    updateGHPAGES "-beta"
fi

#-----------------------------------------------------------------
# MASTER DEPLOY
# - send to google play like release
# - create a changelog
# - add changes to repository
# - commit and push
# - send CHANGELOG.md to gh-pages branch
#-----------------------------------------------------------------
if [[ "$TRAVIS_BRANCH" == "master" && "$TRAVIS_PULL_REQUEST" == "false" && "$TRAVIS_RUN" == "true" ]]; then

    # decrypt deploy on google play file
    openssl aes-256-cbc -K $encrypted_27dcfd0dda78_key -iv $encrypted_27dcfd0dda78_iv -in gplay.tar.gz.enc -out ci/gplay.tar.gz -d

    # uncompress cert file
    tar -zxvf ci/gplay.tar.gz -C ci/

    # sign and deploy to store with fastlane
    fastlane android playstore storepass:'$KEYSTORE' keypass:'$ALIAS'

    # push tag to github
    conventional-github-releaser -t $GH_TOKEN -r 0

    # config git
    configGit
    #------------------------ UPDATE CHANGES --------------------------

    # add modified and delete files
    git add -u

    # create commit
    git commit -m "ci(build): release **version** ${GIT_TAG}"

    # push to branch
    git push origin $TRAVIS_BRANCH

    #------------------------ GH-PAGES --------------------------

    updateGHPAGES
fi